### PR TITLE
Fix #567: Add JsDoc

### LINKED ISSUE #567: JSDoc

### PR DESCRIPTION
**Pull request by edgarmueller**
Add JsDoc to all relevant exported types.

### CODE REVIEW

File: src/core/data.service.ts [hunk 1]
```diff
@@ -1,14 +1,44 @@
...
+/**
+ * A listener that is notified whenever the underlying data changes.
+ */
 export interface DataChangeListener {
+
+  /**
+   * Determines whether this listener is interested in any data change.
+   * If this returns true, the notifyChange method of the listener will be called.
+   * @param {ControlElement} uischema the control element that is affected by the data change
+   */
```
Comment (eneufeld) [inline src/core/data.service.ts]: missing returns annotation

File: src/core/data.service.ts [hunk 2]
```diff
@@ -1,14 +1,44 @@
...
   notifyChange(uischema: ControlElement, newValue: any, data: any): void;
 }
+
+/**
+ * The data service that holds the data of a form and maintains
+ * a list of listeners that are notified whenever the data changes.
+ */
 export class DataService {
   private changeListeners: Array<DataChangeListener>= [];
   constructor(private data: any) {
```
Comment (eneufeld) [inline src/core/data.service.ts]: Constructor documentation

File: src/core/data.service.ts [hunk 3]
```diff
@@ -21,19 +51,40 @@ export class DataService {
...
+  /**
+   * Un-register the given data change listener.
+   * @param {DataChangeListener} listener the data change listener to be un-registered
+   */
   unregisterChangeListener(listener: DataChangeListener): void {
     this.changeListeners.splice(this.changeListeners.indexOf(listener), 1);
   }
+
+  /**
+   * Resolve the ref of the given control agsinst the root data.
```
Comment (eneufeld) [inline src/core/data.service.ts]: agsinst -> against

File: src/core/uischema.registry.ts [hunk 4]
```diff
@@ -2,35 +2,99 @@ import * as _ from 'lodash';
...
+     */
     getBestUiSchema(schema: JsonSchema, data: any): UISchemaElement;
 }
+
+/**
+ * A tester that returns a priority number when a JSON schema and some data,
+ * which determines how likely the associated UI schema will be used to
+ * render the JSON schema/data. The higher the returned number, the more likely
+ * the associated UI schema will be used.
+ */
```
Comment (eneufeld) [inline src/core/uischema.registry.ts]: params and returns missing

File: src/path.util.ts [hunk 5]
```diff
@@ -1,35 +1,59 @@
...
+ * Resolve the given schema path against the given instance until the last
+ * segment. The returned value allows easy assignment of any new value.
+ *
+ * @example
+ * const pair = getValuePropertyPair(someData, someRef);
+ * pair.instance[pair.property] = someValue;
+ *
+ * @param {any} instance the instance to resolve the path against
+ * @param {string} schemaPath the schema path to be resolved
+ * @returns {{instance: string, property: string}} an object containing
```
Comment (eneufeld) [inline src/path.util.ts]: {{instance: string, property: string}} -> {instance: string, property: string}

Comment (edgarmueller) [inline src/path.util.ts]: The double curly braces denote that an object is returned here, so this should be fine.

**Pull request merged**